### PR TITLE
8290079: Reduce interaction with volatile in static initializer of BigInteger

### PR DESCRIPTION
**Pull request by stsypanov**
`BigInteger.powerCache` is volatile and should be assigned only once in static initializer.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290079](https://bugs.openjdk.org/browse/JDK-8290079): Reduce interaction with volatile in static initializer of BigInteger

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [15971c4c](https://git.openjdk.org/jdk/pull/9451/files/15971c4c7a9e7a6b4b94c86cb0e926f35c6f5570)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9451/head:pull/9451` \
`$ git checkout pull/9451`

Update a local copy of the PR: \
`$ git checkout pull/9451` \
`$ git pull https://git.openjdk.org/jdk pull/9451/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9451`

View PR using the GUI difftool: \
`$ git pr show -t 9451`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9451.diff">https://git.openjdk.org/jdk/pull/9451.diff</a>

</details>

### CODE REVIEW

**Review by jaikiran — APPROVED** [review]
The change looks fine to me. Please wait for at least another review from someone more familiar with this area, before merging.

**Comment by rgiulietti** [comment]
(Not a reviewer) While the change doesn't hurt, I doubt that access to a (static) volatile in a static initializer is ever contended.

**Comment by stsypanov** [comment]
@rgiulietti AFAIK volatile access is more expensive than plain one regardless contention

**Comment by rgiulietti** [comment]
Usually yes, but since a static initializer is executed by at most one thread by using a locking protocol before any other static code is ever executed, the runtime _could_ (but I'm not sure it it really does) treat the volatile in the for loop as a local.
But I would approve your change because it makes this more explicit.

**Comment by stsypanov** [comment]
@rgiulietti I've copy-pasted class-loading benchmark from JMH samples
```java
@State(Scope.Thread)
@Warmup(iterations = 10, time = 1)
@Measurement(iterations = 10, time = 5)
@Fork(value = 5, jvmArgsAppend = {"-Xms1g", "-Xmx1g"})
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class Classy {

    @Benchmark
    public Class<?> load() throws ClassNotFoundException {
        return Class.forName("com.tsypanov.slp.Sample", true, new XLoader());
    }

    public static class XLoader extends URLClassLoader {

        private static final byte[] X_BYTECODE = new byte[]{ /*..*/};

        public XLoader() {
            super(new URL[0], ClassLoader.getSystemClassLoader());
        }

        @Override
        protected Class<?> findClass(final String name) {
            return defineClass(name, X_BYTECODE, 0, X_BYTECODE.length);
        }
    }
}
```
and used it to measure loading a class from `byte[]` copied from the following class compiled:
```java
class Sample {
  static volatile int[] items = new int[100];
  static {
    for (int i = 0; i < items.length; i++) {
      items[i] = ThreadLocalRandom.current().nextInt();
    }
  }
}
```
I ran the benchmark with `java -jar target/sleep-benchmarks.jar Classy -prof cl` and for the version above got these results (Java 17):
```
Benchmark                       Mode  Cnt       Score      Error        Units
Classy.load                     avgt   50   96247.202 ±  548.137        ns/op
Classy.load:·class.load         avgt   50  259912.089 ± 1485.314  classes/sec
Classy.load:·class.load.norm    avgt   50       1.000 ±    0.001   classes/op
Classy.load:·class.unload       avgt   50  260243.318 ± 3673.515  classes/sec
Classy.load:·class.unload.norm  avgt   50       1.001 ±    0.014   classes/op
```
Then I've modified the class in the same way I did in this PR:
```java
class Sample {
  static volatile int[] items;
  static {
    int[] items = new int[100];
    for (int i = 0; i < items.length; i++) {
      items[i] = ThreadLocalRandom.current().nextInt();
    }
    Sample.items = items;
  }
}
```
and for modified code got
```

Benchmark                       Mode  Cnt       Score      Error        Units
Classy.load                     avgt   50   63955.673 ±  147.470        ns/op
Classy.load:·class.load         avgt   50  391101.854 ±  925.013  classes/sec
Classy.load:·class.load.norm    avgt   50       1.000 ±    0.001   classes/op
Classy.load:·class.unload       avgt   50  390800.851 ± 2307.589  classes/sec
Classy.load:·class.unload.norm  avgt   50       0.999 ±    0.006   classes/op
```
From this I conclude that volatile costs are still there no matter whether we deal with static or non-static initializers.

**Comment by rgiulietti** [comment]
@stsypanov Hi Сергей, thanks for the convincing measurements!

File: src/java.base/share/classes/java/math/BigInteger.java
Comment (RogerRiggs) [inline src/java.base/share/classes/java/math/BigInteger.java:1276]: Using a different local name will avoid the shadowing and make it more obvious that the initialization is being deliberately being done before the assignment to a volatile.

Comment (stsypanov) [reply src/java.base/share/classes/java/math/BigInteger.java:1276]: Fixed.

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Review by jddarcy — APPROVED** [review]
Version with renamed local variable looks fine.

**Pull request closed without merging**
